File: java/300/LeetCode0350.java
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;

public class LeetCode0350 {
    public int[] intersect(int[] nums1, int[] nums2) {
        if (nums1.length > nums2.length) {
            return intersect(nums2, nums1);
        }
        if (nums1.length == 0) {
            return new int[]{};
        }
        HashMap<Integer,Integer> hii = new HashMap<>();
        for (int num : nums1) {
            hii.put(num, hii.getOrDefault(num, 0) + 1);
        }
        List<Integer> li = new ArrayList<>(nums1.length);
        for (int num : nums2) {
            if (hii.containsKey(num) && hii.get(num) > 0) {
                li.add(num);
                hii.put(num, hii.get(num) - 1);
            }
        }
        int[] res = new int[li.size()];
        int i = 0;
        for (int n : li) {
            res[i] = n;
            ++i;
        }
        return res;
    }
}

File: cpp/L300/LeetCode0350.cc
#include <array>
#include <vector>
using std::vector;

class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        vector<int> res;
        std::array<int, 1001> n1{0}, n2{0};
        for (int n : nums1) {
            n1[n] += 1;
        }
        for (int n : nums2) {
            n2[n] += 1;
        }
        for (int i = 0;i <= 1000;++i) {
            int x = std::min(n1[i], n2[i]);
            if (x > 0) {
                for (int j = 0;j < x;++j) {
                    res.push_back(i);
                }
            }
        }
        return res;
    }
};